tests/common/runner.rs

```diff
@@ -50,30 +53,100 @@ pub fn test_execute_revm<S: Storage + Clone + Send + Sync>(
         false, // TODO: Parameterize this
     );
 }
 
+fn calculate_receipt_root(receipt_envelopes: Vec<ReceiptEnvelope>) -> B256 {
+    // create a BTreeMap (order must be ascending)
+    let entries = receipt_envelopes
+        .into_iter()
+        .enumerate()
+        .map(|(index, receipt)| {
+            let key_buffer = alloy_rlp::encode_fixed_size(&index);
+            let mut value_buffer = Vec::new();
+            receipt.encode_2718(&mut value_buffer);
+            (key_buffer, value_buffer)
+        })
+        .collect::<BTreeMap<_, _>>();
+
+    // calculate the MPT root
+    let mut hash_builder = alloy_trie::HashBuilder::default();
+    for (k, v) in entries {
+        hash_builder.add_leaf(alloy_trie::Nibbles::unpack(&k), &v);
+    }
+    hash_builder.root()
+}
+
+fn get_receipt_envelopes(
+    block: &Block,
+    tx_results: &[PevmTxExecutionResult],
+) -> Vec<ReceiptEnvelope> {
+    // get the receipts
+    let receipt_iter = tx_results.iter().map(|tx| tx.receipt.clone());
+
+    // get the tx_type list
+    let tx_type_iter = block
+        .transactions
+        .txns()
+        .unwrap()
+        .map(|tx| TxType::try_from(tx.transaction_type.unwrap_or_default()).unwrap());
+
+    // zip them to get the receipt envelopes
+    let receipt_envelope_iter =
+        Iterator::zip(receipt_iter, tx_type_iter).map(|(receipt, tx_type)| match tx_type {
+            TxType::Legacy => ReceiptEnvelope::Legacy(receipt.with_bloom()),
+            TxType::Eip2930 => ReceiptEnvelope::Eip2930(receipt.with_bloom()),
+            TxType::Eip1559 => ReceiptEnvelope::Eip1559(receipt.with_bloom()),
+            TxType::Eip4844 => ReceiptEnvelope::Eip4844(receipt.with_bloom()),
+        });
+    receipt_envelope_iter.collect::<Vec<_>>()
+}
+
 // Execute an Alloy block sequentially & with PEVM and assert that
 // the execution results match.
 pub fn test_execute_alloy<S: Storage + Clone + Send + Sync>(
     storage: S,
     block: Block,
     parent_header: Option<Header>,
     must_succeed: bool,
+    must_check_receipts_root: bool,
 ) {
     let concurrency_level = thread::available_parallelism().unwrap_or(NonZeroUsize::MIN);
-    assert_execution_result(
-        pevm::execute(
-            storage.clone(),
-            block.clone(),
-            parent_header.clone(),
-            concurrency_level,
-            true,
-        ),
-        pevm::execute(storage, block, parent_header, concurrency_level, false),
-        must_succeed,
+    let sequential_result = pevm::execute(
+        storage.clone(),
+        block.clone(),
+        parent_header.clone(),
+        concurrency_level,
+        true,
     );
+    let parallel_result = pevm::execute(
+        storage.clone(),
+        block.clone(),
+        parent_header.clone(),
+        concurrency_level,
+        false,
+    );
+    assert_execution_result(&sequential_result, &parallel_result, must_succeed);
+
+    if must_check_receipts_root {
```

Comment:
I don't mind having extra information in code, but this is not that interesting to Parallel EVM, and can be shortened to something like:
```rust
    // We can only calculate the receipts root from Byzantium.
    // Before EIP-658 (https://eips.ethereum.org/EIPS/eip-658), the
    // receipt root is calculated with the post transaction state root,
    // which we doesn't have in these tests.
    if must_match_receipts_root && block.header.number.unwrap() >= 4370000 {
        assert_eq!(
            block.header.receipts_root,
            calculate_receipt_root(get_receipt_envelopes(&block, &sequential_result.unwrap()))
        );
    }
```

Also, looking at this, we can call `get_receipt_envelopes` from inside `calculate_receipt_root` instead of chaining it here.

Reply:
No, the two well-formed functions/steps can still be kept if you move the call to `build_receipt_envelopes` into `calculate_receipt_root`. The point is the caller doesn't need to know that, either in test modules or when we add it to the library's API. This is like asking PEVM user to call `execute(prepropcess(&block))` instead of calling `execute(&block)`.